week7_02.py

```diff
@@ -0,0 +1,123 @@
+class PostOffice:
+    """A Post Office class. Allows users to message each other.
+
+    Args:
+        usernames (list): Users for which we should create PO Boxes.
+
+    Attributes:
+        message_id (int): Incremental id of the last message sent.
+        boxes (dict): Users' inboxes.
+    """
+
+    def __init__(self, usernames):
+        self.message_id = 0
+        self.boxes = {user: [] for user in usernames}
+
+    def send_message(self, sender, recipient, message_body, urgent=False):
+        """Send a message to a recipient.
+
+        Args:
+            sender (str): The message sender's username.
+            recipient (str): The message recipient's username.
+            message_body (str): The body of the message.
+            urgent (bool, optional): The urgency of the message.
+                                    Urgent messages appear first.
+
+        Returns:
+            int: The message ID, auto incremented number.
+
+        Raises:
+            KeyError: If the recipient does not exist.
+
+        Examples:
+            After creating a PO box and sending a letter,
+            the recipient should have 1 message in the
+            inbox.
+
+            po_box = PostOffice(['a', 'b'])
+            message_id = po_box.send_message('a', 'b', 'Hello!')
+            len(po_box.boxes['b'])
+            1
+            message_id
+            1
+        """
+        user_box = self.boxes[recipient]
+        self.message_id = self.message_id + 1
+        message_details = {
+            'id': self.message_id,
+            'body': message_body,
+            'sender': sender,
+            # Add bool to mark if read or not
+            'read': False
+        }
+        if urgent:
+            user_box.insert(0, message_details)
+        else:
+            user_box.append(message_details)
+        return self.message_id
+
+    def read_inbox(self, usernames: str, num_of_messages_to_read: int = None):
+        """
+        Return the first "num_of_messages_to_read" messages in the user's inbox.
+        If no "num_of_messages_to_read" have been forwarded, return all messages in the user's inbox.
+        The messages will be marked as read and will not be returned to the user on the next call.
+        :param usernames: Username you want to read messages
+        :param num_of_messages_to_read: (int) The maximum number of messages the user wants to read
+        :return: list_of_messages_to_return List of "num_of_messages_to_read" which have not yet been read
+        """
+
+        list_of_messages_to_return = []
+        if not num_of_messages_to_read or num_of_messages_to_read > len(self.boxes[usernames]):
+            num_of_messages_to_read = len(self.boxes[usernames])
+
+        # It could have been written shorter but it came at the expense of understanding the code
+        for message in self.boxes[usernames][:num_of_messages_to_read]:
```

Comment:
I fould have used filter, to filter the ones where read = False, and then a map . 
By the way, In larger code, I would do this https://stackoverflow.com/questions/1080026/is-there-a-map-without-result-in-python